presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -174,20 +175,113 @@ public static Block geometryToBingTiles(@SqlType(GEOMETRY_TYPE_NAME) Slice input
+        }
+
+        int minZoomLevelForOptimizedTiling = 10;
 
-                if (pointOrRectangle || !tileToPolygon(tile).disjoint(geometry)) {
-                    BIGINT.writeLong(blockBuilder, tile.encode());
+        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), tileCount);
+        if (pointOrRectangle || zoomLevel <= minZoomLevelForOptimizedTiling) {
+            // Collect tiles covering bounding box and check each tile for intersection with the geometry.
+            // Skip intersection check if geometry is a point or rectangle. In these cases, by definition,
+            // all tiles covering bounding box intersect the geometry.
+            for (int x = leftUpperTile.getX(); x <= rightLowerTile.getX(); x++) {
+                for (int y = leftUpperTile.getY(); y <= rightLowerTile.getY(); y++) {
+                    BingTile tile = BingTile.fromCoordinates(x, y, zoomLevel);
+                    if (pointOrRectangle || !tileToPolygon(tile).disjoint(geometry)) {
+                        BIGINT.writeLong(blockBuilder, tile.encode());
+                    }
+                }
+            }
+        }
+        else {
+            // Intersection checks above are expensive. The logic below attempts to reduce the number
+            // of these checks. The idea is to identify large tiles which are fully covered by the
+            // geometry. For each such tile, we can cheaply compute all the containing tiles at
+            // the right zoom level and append them to results in bulk. This way we perform a single
+            // containment check instead of 2 to the power of level delta intersection checks, where
+            // level delta is the difference between the desired zoom level and level of the large
+            // tile covered by the geometry.
+            int divider = 1 << (zoomLevel - minZoomLevelForOptimizedTiling);
+            int minX = (int) Math.floor(leftUpperTile.getX() / divider);
+            int maxX = (int) Math.floor(rightLowerTile.getX() / divider);
+            int minY = (int) Math.floor(leftUpperTile.getY() / divider);
+            int maxY = (int) Math.floor(rightLowerTile.getY() / divider);
+
+            for (int x = minX; x <= maxX; x++) {
+                for (int y = minY; y <= maxY; y++) {
+                    BingTile tile = BingTile.fromCoordinates(x, y, minZoomLevelForOptimizedTiling);
+                    geometryToBingTiles(geometry, zoomLevel, tile, blockBuilder);
                 }
             }
         }
 
         return blockBuilder.build();
     }
 
+    /**
+     * Identifies a minimum set of tiles at specified zoom level that cover intersection of the
+     * specified geometry and a specified tile of the same or lower level. Adds tiles to provided
+     * BlockBuilder.
+     *
+     * @pre tile.getZoomLevel() <= zoomLevel
+     */
+    private static void geometryToBingTiles(OGCGeometry geometry,
```

Comment:
the name of this method sounds like it should return bing tiles, so we should rename it to sth like `writeTilesToBlockBuilder`